core/src/com/unciv/logic/map/UnitMovementAlgorithms.kt

```diff
@@ -425,12 +427,27 @@ class UnitMovementAlgorithms(val unit:MapUnit) {
                 // Anyway: PANIC!! We stop this route, and after leaving the game in a valid state,
                 // we try again.
                 needToFindNewRoute = true
-                break
+                break // If you ever remove this break, remove the `assumeCanPassThrough` param below
             }
-            unit.removeFromTile()
-            unit.putInTile(tile)
+            unit.moveThroughTile(tile)
+            
+            // In case something goes wrong, cache the last tile we were able to end on
+            // We can assume we can pass through this tile, as we would have broken earlier
+            if (unit.movement.canMoveTo(tile, assumeCanPassThrough = true)) {
+                lastReachedEnterableTile = tile
+            }
+            
             if (unit.isDestroyed) break
         }
+        
+        if (!unit.isDestroyed)
+            unit.putInTile(lastReachedEnterableTile)
+
+        if (!unit.civInfo.gameInfo.gameParameters.godMode) {
+            unit.currentMovement -= distanceToTiles[lastReachedEnterableTile]!!.totalDistance
+            if (unit.currentMovement < 0.1) unit.currentMovement = 0f // silly floats which are "almost zero"
```

Comment:
0.1 is a Double, and the operator will lift its other operand to Double too before comparing I think - and I would use an epsilon of `0.05f` as 0.1 is the cost of 1 tile along a railroad.

Reply:
Not my code, just copy-pasted from somewhere else :) That being said, I do agree this should be 0.05f